### DISCUSSION TITLE
posts/flex-consumption-plan

### DISCUSSION THREAD

**Discussion by giscus** [TechWatching/techwatching.dev#62] [Announcements]
# posts/flex-consumption-plan

In this article, we will explore how to provision a Function App in the new Azure Functions hosting plan: the Flex Consumption plan. We will do that using Pulumi and TypeScript.

https://techwatching.dev/posts/flex-consumption-plan

**Comment by mnbf9rca** [comment]
Do you know if it's possible to deploy the function app without using function core tools? For a regular consumption plan, we can create the zip file ourselves and push it, but i can't get that working for consumption flex.

**Reply by TechWatching** [reply]
That's a good question but I don't know the answer.  You could probably push the zip yourself but I think you would have to call an Azure API anyway for the function app to retrieve this package and sync its triggers.

Why don't you want to use the core tools by the way?

**Reply by mnbf9rca** [reply]
i want my deployments to be deterministic, and that means deploying a single build artefact, not recreating it for each deployment. Also i prefer to have fewer dependencies, so if i can get pulumi to just handle the deployment of the zip file (like you can with regular consumption plan) that would be great - but it doesnt work.

**Reply by ddempsey-ignite** [reply]
Hey @mnbf9rca, did you ever get this working? I'm currently in a same situation to you. I've deployed a function app and related infrastructure via pulumi and have it pointing to a dedicated storage account/container. A zip file of function code has been uploaded there, however, it just seems unable to get the code and deploy the function.

**Reply by mnbf9rca** [reply]
No. Instead i created a bash script. I output `function_app_name` and `content_changed` as stack outputs and use the function app core tool cli to deploy the actual function app.
```
#!/bin/bash

# Check if environment argument is provided
if [ -z "$1" ]; then
  echo "Environment parameter is missing."
  echo "Usage: $0 <env> [--always-deploy]"
  exit 1
fi

ENV=$1
ALWAYS_DEPLOY=false

# Check for optional --always-deploy argument
if [ "$2" == "--always-deploy" ]; then
  ALWAYS_DEPLOY=true
fi

# Check if az cli is logged in
if ! az account show > /dev/null 2>&1; then
  echo "Error: Azure CLI is not logged in. Please log in using 'az login' before running the script."
  exit 1
fi

# Check if host.json exists in the current working directory
if [ ! -f "./host.json" ]; then
  echo "Error: host.json file is missing in the current directory. Function app deployment requires host.json file."
  exit 1
fi

# Run pulumi up with the specified environment
pulumi stack select --stack "$ENV" --non-interactive
pulumi up --yes --non-interactive

# Exit if pulumi up is not successful
if [ $? -ne 0 ]; then
  echo "Error: pulumi up command failed. Exiting."
  exit 1
fi

# Retry loop to check if the function app exists
MAX_RETRIES=10
RETRY_DELAY=15
APP_NAME=$(pulumi stack output function_app_name)

for ((i=1; i<=MAX_RETRIES; i++)); do
  echo "Checking if function app '$APP_NAME' exists (attempt $i)..."
  if az functionapp show --name "$APP_NAME" --resource-group "$(pulumi stack output resource_group_name)" > /dev/null 2>&1; then
    echo "Function app '$APP_NAME' found."
    break
  else
    echo "Function app '$APP_NAME' not found. Retrying in $RETRY_DELAY seconds..."
    sleep "$RETRY_DELAY"
  fi
  
  if [ $i -eq $MAX_RETRIES ]; then
    echo "Function app '$APP_NAME' still not found after $MAX_RETRIES attempts. Exiting."
    exit 1
  fi
done

# Publish the function app if pulumi up command was successful and one of the following is true:
# - content_changed output is true
# - --always-deploy argument is provided
if [ "$(pulumi stack output content_changed)" == "true" ] || [ "$ALWAYS_DEPLOY" == true ]; then
  # Publish the function app using function core tools
  func azure functionapp publish "$APP_NAME" --verbose --python
else
  echo "Skipping function app publish as content did not change, or --always-deploy was not used."
fi
```

**Reply by TechWatching** [reply]
Concerning building a single artifact and deploying it to different environments, it's easy to do with the CLI:

```bash
dotnet publish 
cd bin/release/net8. 0/
func azure functionapp publish "func-myfunctionname" --no-build
```

If you want to use Pulumi to handle the code deployment, I don't think there is a built-in way to do that because FlexConsumption use [OneDeploy](https://learn.microsoft.com/en-us/azure/azure-functions/functions-deployment-technologies?tabs=windows#one-deploy) and not zip deploy nor external package url.
